File: linked_list/MiddleElement.cpp
#include<iostream>
using namespace std;

class Node
{
public:
	int data;
	class Node *next;
	Node(int data)
	{
		this->data = data;
		next=NULL;
	}
};

void push(class Node **head, int data)
{
	class Node *node = new Node(data);
	node->next = *head;
	*head = node;
}

void print_list(class Node *head)
{
	while(head!=NULL)
	{
		cout<<head->data<<" -> ";
		head = head->next;
	}
	cout<<"NULL"<<endl;
}

void print_middle(class Node *head)
 {
 	if (head == NULL)
 		{
 			cout<<"NULL"<<endl;
 			return;
 		}	

 	class Node *fast_ptr = head;
 	class Node *slow_ptr = head;	
 		
 	while(fast_ptr!=NULL)
 	{
 		if (fast_ptr->next==NULL)
			break;
  		fast_ptr = fast_ptr->next->next;
 		slow_ptr = slow_ptr->next;
 	}
 	cout<<"Middle Element is "<< slow_ptr->data<<endl;
 }
int main()
{
	class Node *head = NULL;

	int n, data;
	cout<<"Enter Length of Linked list: ";
	cin>>n;
	for (int i=0; i<n; i++)
	{
		cout<<"Data: ";
		cin>>data;
		push(&head, data);
	}

	print_list(head);

	print_middle(head);
	return 0;
}

File: linked_list/MiddleElement.java
import java.util.*;
class MiddleElement
{
	class Node
	{
		int data;
		Node next;
		Node(int data)
		{
			this.data = data;
			next = null;
		}
	}
	Node head;
	public void push(int data)
	{
		Node node = new Node(data);
		node.next = head;
		head = node;
	}

	public void show()
	{
		Node ptr = head;
		while(ptr!=null)
		{
			System.out.print(ptr.data+" -> ");
			ptr = ptr.next;
		}
		System.out.println("null");
	}

	public void print_middle()
	{
		if(head==null)
		{
			System.out.println("Linked List null");
			return;
		}
		Node slow_ptr = head;
		Node fast_ptr = head;

		while(fast_ptr!=null) // For Even Length
		{
			if (fast_ptr.next!=null) // For Odd Length
			{
				fast_ptr = fast_ptr.next.next;
				slow_ptr = slow_ptr.next;
			}	
			else
				break;
		}
		System.out.println("Middle Element is: "+ slow_ptr.data);
	}
	public static void main(String[] args)
	{
		MiddleElement list = new MiddleElement();

		System.out.print("Enter the length of Linked List: ");
		Scanner sc = new Scanner(System.in);
		
		int n = sc.nextInt();

		for(int i=0; i<n; i++)
		{
			System.out.print("Enter Element: ");
			int data = sc.nextInt();
			list.push(data);
		}

		list.show();
		list.print_middle();
	}
}	